ambry-account/src/main/java/com/github/ambry/account/HelixAccountService.java

```diff
@@ -544,4 +531,127 @@ private boolean containsName(String name) {
+      this.updatedAccounts = updatedAccounts;
+    }
+
+    @Override
+    public ZNRecord update(ZNRecord currentData) {
+      ZNRecord newRecord;
+      if (currentData == null) {
+        logger.debug(
+            "ZNRecord does not exist on path={} in HelixPropertyStore when updating accounts. Creating a new ZNRecord.",
+            FULL_ACCOUNT_METADATA_PATH);
+        newRecord = new ZNRecord(ZN_RECORD_ID);
+      } else {
+        newRecord = currentData;
+      }
+      Map<String, String> accountMap = newRecord.getMapField(ACCOUNT_METADATA_MAP_KEY);
+      if (accountMap == null) {
+        logger.debug("AccountMap does not exist in ZNRecord when updating accounts. Creating a new accountMap");
+        accountMap = new HashMap<>();
+      }
+      try {
+        remoteAccountInfoMap = new AccountInfoMap(accountMap);
+      } catch (JSONException e) {
+        // Do not depend on Helix to log, so log the error message here.
+        logger.error("Exception occurred when building AccountInfoMap from accountMap={}", accountMap, e);
+        accountServiceMetrics.remoteDataCorruptionErrorCount.inc();
+        throw new IllegalStateException("Exception occurred when building AccountInfoMap from accountMap", e);
+      }
+      // if there is any conflict with the existing record, fail the update. Exception thrown in this updater will
+      // be caught by Helix and helixStore#update will return false.
+      if (hasConflictingAccount(updatedAccounts, remoteAccountInfoMap)) {
+        // Throw exception, so that helixStore can capture and terminate the update operation
+        throw new IllegalArgumentException(
+            "Updating accounts failed because one account to update conflicts with existing updatedAccounts");
+      } else {
+        for (Account account : updatedAccounts) {
+          try {
+            accountMap.put(String.valueOf(account.getId()), account.toJson().toString());
+          } catch (Exception e) {
+            String message = "Updating accounts failed because unexpected exception occurred when updating accountId="
+                + account.getId() + " accountName=" + account.getName();
+            // Do not depend on Helix to log, so log the error message here.
+            logger.error(message, e);
+            throw new IllegalStateException(message, e);
+          }
+        }
+        newRecord.setMapField(ACCOUNT_METADATA_MAP_KEY, accountMap);
+        return newRecord;
+      }
+    }
+
+    /**
+     * Save the state from zookeeper from before the update and after the update to disk.
+     * This will write the backup content string to the disk with the following file name format:
+     * {@code {yyyyMMdd}T{HHmmss}-{unique long}.bak}.
+     * If there are multiple files with the same timestamp, the unique long will prevent the file names from clashing.
+     * @param succeeded {@code true} iff the update succeeded.
+     */
+    void persistBackup(boolean succeeded) {
```

Comment:
How frequently is this method expected to run?

Reply:
This is called whenever someone makes changes to their account (like creating a new container). We expect this to happen a few times an hour, max.